README
NAME
    RT-Extension-Tags - Provides simple tagging using custom fields

DESCRIPTION
    This extension allows you to create tags using custom fields on tickets.
    It adds a new custom field type, "Tags", which allows users to add new
    values (tags) that will then be added to the list of available
    autocomplete values for that custom field.

    The created tags become links to a search of all active tickets with
    that tag.

  Tag Custom Field
    The initdb step installs an example global Tag custom field.

RT VERSION
    Works with RT 4.2, 4.4, 5.0

INSTALLATION
    perl Makefile.PL
    make
    make install
        May need root permissions

    Edit your /opt/rt4/etc/RT_SiteConfig.pm
        If you are using RT 4.2 or greater, add this line:

            Plugin('RT::Extension::Tags');

        For RT 4.0, add this line:

            Set(@Plugins, qw(RT::Extension::Tags));

        or add RT::Extension::Tags to your existing @Plugins line.

    make initdb
        This optional step installs an example global Tag custom field.

    Clear your mason cache
            rm -rf /opt/rt4/var/mason_data/obj

    Restart your webserver

  UPGRADING
    If you are upgrading from version 0.01 or 0.02, note that the custom
    field type was changed from a default RT "multiple values with
    autocompletion" to a dedicated tag custom field type. To upgrade:

    *   Change your Tag custom field to use the new tag custom field type

        You can either edit your existing custom field and change the Type
        to "Enter multiple tags" or run the initdb step and copy your values
        to the new Tag custom field automatically created.

    *   Disable the scrip "On Custom Field Change Add New Tag Values"

        This scrip is no longer needed with the new tag custom field type,
        so you can disable or delete it.

AUTHOR
    Best Practical Solutions, LLC <modules@bestpractical.com>

BUGS
    All bugs should be reported via email to

        L<bug-RT-Extension-Tags@rt.cpan.org|mailto:bug-RT-Extension-Tags@rt.cpan.org>

    or via the web at

        L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-Tags>.

LICENSE AND COPYRIGHT
    This software is Copyright (c) 2016 by Best Practical Solutions, LLC

    This is free software, licensed under:

      The GNU General Public License, Version 2, June 1991



Changes
Revision history for RT-Extension-Tags

0.05 2020-07-07
 - Update documentation & Makefile.PL for RT 5 release

0.04 2017-02-20

 - Update tag input elements to be consistent with usage in RT 4.2, 4.4

0.03 2016-08-08

 - Create a new RT custom field type for tags to use in place of
   RT's multiple select type. Add javascript tagit library to enhance
   UI when selecting tags.

   Note: the CF type change requires manual updates if you installed
   previous versions. See the Upgrade notes for details.

0.02 2016-07-13

 - Contributions from Alex Vandiver to remove the custom condition,
   remove an incorrect error check, refactor the tag search query
   generation, and various other clean-ups.

0.01 2016-07-11
 - Initial version


lib/RT/Action/AddNewCFValue.pm
package RT::Action::AddNewCFValue;

use strict;
use warnings;

use base qw(RT::Action);

sub CustomField {
    my $self    = shift;
    my $cf_name = $self->Argument;

    my $cf = $self->TicketObj->LoadCustomFieldByIdentifier($cf_name);
    RT::Logger->error("Unable to load custom field $cf_name") unless $cf->Id;
    return $cf;
}

sub Prepare {
    my $self = shift;
    my $txn  = $self->TransactionObj;
    my $cf   = $self->CustomField;

    return 1 if $txn->Type eq "Create";
    return 1 if $txn->Type eq "CustomField" and $txn->Field == $cf->id;
    return 0;
}

=head2 DESCRIPTION

If the current value provided for this custom field isn't currently in
the list of available values, add it.

The custom field name is identified by the Argument set for the scrip
action when the action is installed.

=cut

sub Commit {
    my $self = shift;
    my $txn  = $self->TransactionObj;
    my $cf   = $self->CustomField;

    my $values_obj = $cf->Values;
    my @current_values;
    while ( my $cf_value = $values_obj->Next ) {
        push @current_values, $cf_value->Name;
    }

    if ( $txn->Type eq 'Create' ) {

        # Can get multiple new values from create, so check them all
        my @ticket_values = split /\n/,
            $self->TicketObj->CustomFieldValuesAsString($cf);

        foreach my $value (@ticket_values) {

            # Already got it
            next if grep { $value eq $_ } @current_values;

            # Don't have it, so add it
            $cf->AddValue( Name => $value );
        }
        return 1;
    } else {

        # Defer to core's NewReferenceObject if possible
        my $cf_value = $txn->can("NewReferenceObject")
            ? $txn->NewReferenceObject
            : $self->NewReferenceObject($txn);

        return 0 unless $cf_value and $cf_value->Id;

        foreach my $value (@current_values) {

            # Already have it
            return 1 if $cf_value->Content eq $value;
        }

        # It's new, add it
        $cf->AddValue( Name => $cf_value->Content );
        return 1;
    }
}

# Lifted from RT 4.2's RT::Transaction, for compatibility with RT
# 4.0.  Can be removed after 4.0 support drops.

sub NewReferenceObject {
    my $self = shift;
    my $txn  = shift;
    my $type = $txn->__Value("ReferenceType");
    my $id   = $txn->__Value("NewReference");
    return unless $type and $id;

    my $object = $type->new( $self->CurrentUser );
    $object->Load($id);
    return $object;
}

1;


lib/RT/Extension/Tags.pm
use strict;
use warnings;
package RT::Extension::Tags;

our $VERSION = '0.05';


require RT::CustomField;

$RT::CustomField::FieldTypes{Tags} = {
    sort_order     => 85,
    selection_type => 1,
    labels         => [
        'Enter multiple tags',  # loc
        'Enter one tag',        # loc
        'Enter up to [_1] tag', # loc
    ],
};

RT->AddJavaScript("tag-it.min.js");
RT->AddStyleSheets("jquery.tagit.css");

no warnings 'redefine';
my $old = \&RT::CustomField::AddValueForObject;
*RT::CustomField::AddValueForObject = sub {
    my $self = shift;
    my %args = (
        Content => undef,
        LargeContent => undef,
        @_
    );

    my ($ok, $msg) = $old->($self, @_);
    return ($ok, $msg) unless $ok;

    return ($ok, $msg) unless $self->Type eq "Tags";


    my $value = $args{LargeContent} || $args{Content};
    my $as_super = RT::CustomField->new( RT->SystemUser );
    $as_super->Load( $self->id );
    my $values = $as_super->Values;
    $values->Limit( FIELD => 'Name', VALUE => $value );
    return ($ok, $msg) if $values->Count;

    $as_super->AddValue( Name => $value );
    return ($ok, $msg);
};


=head1 NAME

RT-Extension-Tags - Provides simple tagging using custom fields

=head1 DESCRIPTION

This extension allows you to create tags using custom fields on
tickets.  It adds a new custom field type, "Tags", which allows users
to add new values (tags) that will then be added to the list of
available autocomplete values for that custom field.

The created tags become links to a search of all active tickets
with that tag.

=head2 Tag Custom Field

The initdb step installs an example global Tag custom field.

=head1 RT VERSION

Works with RT 4.2, 4.4, 5.0

=head1 INSTALLATION

=over

=item C<perl Makefile.PL>

=item C<make>

=item C<make install>

May need root permissions

=item Edit your F</opt/rt4/etc/RT_SiteConfig.pm>

If you are using RT 4.2 or greater, add this line:

    Plugin('RT::Extension::Tags');

For RT 4.0, add this line:

    Set(@Plugins, qw(RT::Extension::Tags));

or add C<RT::Extension::Tags> to your existing C<@Plugins> line.

=item C<make initdb>

This optional step installs an example global C<Tag> custom field.

=item Clear your mason cache

    rm -rf /opt/rt4/var/mason_data/obj

=item Restart your webserver

=back

=head2 UPGRADING

If you are upgrading from version 0.01 or 0.02, note that the custom field
type was changed from a default RT "multiple values with autocompletion" to
a dedicated tag custom field type. To upgrade:

=over

=item * Change your Tag custom field to use the new tag custom field type

You can either edit your existing custom field and change the Type to "Enter
multiple tags" or run the initdb step and copy your values to the new Tag
custom field automatically created.

=item * Disable the scrip "On Custom Field Change Add New Tag Values"

This scrip is no longer needed with the new tag custom field type,
so you can disable or delete it.

=back

=head1 AUTHOR

Best Practical Solutions, LLC E<lt>modules@bestpractical.comE<gt>

=head1 BUGS

All bugs should be reported via email to

    L<bug-RT-Extension-Tags@rt.cpan.org|mailto:bug-RT-Extension-Tags@rt.cpan.org>

or via the web at

    L<rt.cpan.org|http://rt.cpan.org/Public/Dist/Display.html?Name=RT-Extension-Tags>.

=head1 LICENSE AND COPYRIGHT

This software is Copyright (c) 2016 by Best Practical Solutions, LLC

This is free software, licensed under:

  The GNU General Public License, Version 2, June 1991

=cut

1;
